File: apps/schemas/tax.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from apps.models.tax import (
    TaxAdjustmentDirection,
    TaxFilingModel,
    TaxPeriodStatus,
    TaxSystem,
)


class TaxConfigurationUpdate(BaseModel):
    tax_system: TaxSystem = TaxSystem.IGIC
    filing_model: TaxFilingModel = TaxFilingModel.MODEL_420
    filing_frequency: str = Field(default="quarterly", pattern="^quarterly$")
    monthly_tracking_enabled: bool = True
    default_tax_rate: Decimal = Field(default=Decimal("7.00"), ge=0, le=100, decimal_places=2)
    notes: str | None = Field(default=None, max_length=3000)

    @model_validator(mode="after")
    def validate_model(self) -> "TaxConfigurationUpdate":
        if self.tax_system == TaxSystem.IGIC and self.filing_model != TaxFilingModel.MODEL_420:
            raise ValueError("El régimen IGIC de esta versión utiliza el modelo 420")
        return self


class TaxConfigurationRead(TaxConfigurationUpdate):
    id: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TaxRateBreakdown(BaseModel):
    tax_rate: Decimal
    taxable_base: Decimal
    tax_amount: Decimal


class TaxAdjustmentCreate(BaseModel):
    year: int = Field(ge=2000, le=2100)
    month: int = Field(ge=1, le=12)
    direction: TaxAdjustmentDirection
    amount: Decimal = Field(gt=0, decimal_places=2)
    description: str = Field(min_length=2, max_length=500)


class TaxAdjustmentRead(BaseModel):
    id: str
    year: int
    month: int
    direction: TaxAdjustmentDirection
    amount: Decimal
    description: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TaxMonthSummary(BaseModel):
    year: int
    month: int
    month_label: str
    status: TaxPeriodStatus
    output_base: Decimal
    output_tax: Decimal
    input_base: Decimal
    input_tax: Decimal
    output_adjustments: Decimal
    input_adjustments: Decimal
    result: Decimal
    result_type: str
    output_breakdown: list[TaxRateBreakdown]
    input_breakdown: list[TaxRateBreakdown]
    adjustments: list[TaxAdjustmentRead]
    closed_at: datetime | None
    legal_notice: str


class TaxPeriodAction(BaseModel):
    notes: str | None = Field(default=None, max_length=2000)


class TaxQuarterSummary(BaseModel):
    year: int
    quarter: int
    period_label: str
    months: list[TaxMonthSummary]
    output_base: Decimal
    output_tax: Decimal
    input_base: Decimal
    input_tax: Decimal
    output_adjustments: Decimal
    input_adjustments: Decimal
    result: Decimal
    result_type: str
    model: str
    filing_window: str
    all_months_closed: bool
    legal_notice: str

File: apps/portal/src/api/tax.ts
import { apiRequest, authenticatedDownload } from './client'

export type TaxSystem = 'igic' | 'iva'
export type TaxPeriodStatus = 'open' | 'closed'
export type TaxAdjustmentDirection = 'output' | 'input'

export interface TaxConfiguration {
  id: string
  tax_system: TaxSystem
  filing_model: '420'
  filing_frequency: 'quarterly'
  monthly_tracking_enabled: boolean
  default_tax_rate: string
  notes: string | null
  created_at: string
  updated_at: string
}

export type TaxConfigurationPayload = Omit<TaxConfiguration, 'id' | 'created_at' | 'updated_at'>

export interface TaxRateBreakdown {
  tax_rate: string
  taxable_base: string
  tax_amount: string
}

export interface TaxAdjustment {
  id: string
  year: number
  month: number
  direction: TaxAdjustmentDirection
  amount: string
  description: string
  created_at: string
}

export interface TaxMonthSummary {
  year: number
  month: number
  month_label: string
  status: TaxPeriodStatus
  output_base: string
  output_tax: string
  input_base: string
  input_tax: string
  output_adjustments: string
  input_adjustments: string
  result: string
  result_type: 'to_pay' | 'to_compensate' | 'zero'
  output_breakdown: TaxRateBreakdown[]
  input_breakdown: TaxRateBreakdown[]
  adjustments: TaxAdjustment[]
  closed_at: string | null
  legal_notice: string
}

export interface TaxQuarterSummary {
  year: number
  quarter: number
  period_label: string
  months: TaxMonthSummary[]
  output_base: string
  output_tax: string
  input_base: string
  input_tax: string
  output_adjustments: string
  input_adjustments: string
  result: string
  result_type: 'to_pay' | 'to_compensate' | 'zero'
  model: string
  filing_window: string
  all_months_closed: boolean
  legal_notice: string
}

export interface TaxAdjustmentPayload {
  year: number
  month: number
  direction: TaxAdjustmentDirection
  amount: string
  description: string
}

export const getTaxConfiguration = () =>
  apiRequest<TaxConfiguration>('/tax/configuration')

export const updateTaxConfiguration = (payload: TaxConfigurationPayload) =>
  apiRequest<TaxConfiguration>('/tax/configuration', {
    method: 'PUT',
    body: JSON.stringify(payload),
  })

export const getTaxMonths = (year: number) =>
  apiRequest<TaxMonthSummary[]>(`/tax/months?year=${year}`)

export const getModel420 = (year: number, quarter: number) =>
  apiRequest<TaxQuarterSummary>(`/tax/model-420/${year}/${quarter}`)

export const closeTaxMonth = (year: number, month: number, notes: string | null = null) =>
  apiRequest<TaxMonthSummary>(`/tax/months/${year}/${month}/close`, {
    method: 'POST',
    body: JSON.stringify({ notes }),
  })

export const reopenTaxMonth = (year: number, month: number) =>
  apiRequest<TaxMonthSummary>(`/tax/months/${year}/${month}/open`, { method: 'POST' })

export const createTaxAdjustment = (payload: TaxAdjustmentPayload) =>
  apiRequest<TaxAdjustment>('/tax/adjustments', {
    method: 'POST',
    body: JSON.stringify(payload),
  })

export const downloadModel420Csv = (year: number, quarter: number) =>
  authenticatedDownload(
    `/tax/model-420/${year}/${quarter}/csv`,
    `modelo-420-${year}-T${quarter}-borrador.csv`,
  )
